### DISCUSSION TITLE
Access OAuth provider specific information

### DISCUSSION THREAD

**Discussion by davidbrochart** [fief-dev/fief#220] [Q&A] [answered]
I'd like to specify GitHub [scopes](https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/scopes-for-oauth-apps#available-scopes) to retrieve more information about a user, and in particular `read:user` in order to read a user's profile data, so I added `read:user` to the list of [scopes](https://docs.fief.dev/configure/oauth-providers/#scopes) for my GitHub OAuth provider in Fief.
But I guess I should allow that in GitHub's "account permissions" for the OAuth app? Unfortunately the only thing I found was "Profile", and I can only grant "Read and write", while I'd only want "Read".
Then I don't know how to access this user information from my server. Is it accessible from Fief's API or should I use the GitHub API?

**Comment by davidbrochart** [comment]
This looks useful: https://docs.github.com/en/rest/users/users?apiVersion=2022-11-28#get-the-authenticated-user

**Reply by davidbrochart** [reply]
Unfortunately I get a `401` with this:
```py
async def foo(
    access_token_info: FiefAccessTokenInfo = Depends(auth.authenticated()),
):
    token = access_token_info["access_token"]
    async with httpx.AsyncClient() as client:
        headers = {
            "Authorization": f"Bearer {token}",
            "Accept": "application/vnd.github+json",
            "X-GitHub-Api-Version": "2022-11-28",
        }
        r = await client.get("https://api.github.com/user", headers=headers)
```

**Comment by frankie567** [comment, marked as answer]
Regarding the scopes, Fief will always by default ask for the scope `user` and `user:email` with GitHub. So you should already be covered to call `https://api.github.com/user`. Apparently, there is no specific permission to enable on the GitHub app settings, apart from "Email addresses" -> "Read-only".

---

Now, as to how to call the GitHub API. What you show here won't work because you try to use the Fief access token on GitHub, which doesn't recognize it.

However, Fief provides an Admin API so you can generate fresh access tokens from the connected OAuth Provider. You can read more about it here: https://docs.fief.dev/going-further/oauth-provider-token/

So basically:

1. Using the Admin API, you ask Fief to give you a valid GitHub access token
2. You call the GitHub API with this token

**Reply by davidbrochart** [reply]
Thanks François, that works!
I have another question: is it possible to know which OAuth provider was used to authenticate a user? Since I want to make provider-specific API requests, I need to know which provider it is.

**Reply by davidbrochart** [reply]
Also, I'd like to update my Fief user profile with information that I get from GitHub in the auth callback endpoint, but I'm not sure how to get the user ID in there. I think at this point the user is authenticated, but I can obviously not depend on the current user yet.

**Reply by frankie567** [reply]
1. Yes, the Admin API can give you the list of associated OAuth Accounts on a given user: [`/admin/api/users/{USER_ID}/oauth-accounts`](https://example.fief.dev/admin/api/docs#/Users/users_list_oauth_accounts_users__id__oauth_accounts_get)
2. When calling `fief.auth_callback`, you get a tuple with two dictionaries: `tokens` and `userinfo`. `userinfo` is a [typed dictionary](https://fief-dev.github.io/fief-python/fief_client.html#FiefUserInfo), where `sub` is the ID of the user, i.e. `userinfo["sub"]`.

**Reply by davidbrochart** [reply]
Thanks so much!

**Reply by davidbrochart** [reply]
1. Although it doesn't tell me which account was used to authenticate the user, right? So how would I go to populate my Fief user account with information from the OAuth provider?

**Reply by frankie567** [reply]
Well, as for now, a user can only have one associated OAuth Account, so the list is either empty or has one account.

In the general case, you can loop through all the accounts and look for the provider you're interested in. You have access to `oauth_provider_id` and a representation of the provider through `oauth_provider`, so you should be able to identify it quite easily.

**Reply by davidbrochart** [reply]
> Well, as for now, a user can only have one associated OAuth Account, so the list is either empty or has one account.

OK, so it's currently easy.
But in the general case, I still don't see how I can for example query GitHub if the user used GitHub to authenticate, or query Google if they used Google. It seems I'm missing this information at runtime.

**Reply by frankie567** [reply]
Indeed, the callback doesn't give hints about the method used to authenticate. I'm not sure if it's something possible to add in an OpenID Connect flow.

That said, imagine we have a user who can use two OAuth Providers (say, GitHub and Google) to sign in (again, that's not possible currently with Fief). When you are in the callback, you look for their OAuth Account. It returns you two objects: one for GitHub, one for Google. Both are valid and can be used to retrieve data both from GitHub and Google. So, I would say that you can just loop through all the accounts and retrieve the information you need.

Another thing: when the user successfully signs in using an OAuth Provider, we update it with the latest access and refresh tokens in Fief's database. So, you can look for the most recent one based on `updated_at`: `/admin/api/users/{USER_ID}/oauth-accounts?ordering=-updated_at` ➡️ the first one in `results` list is the most recently updated

**Reply by davidbrochart** [reply]
> the first one in `results` list is the most recently updated

I'll keep that in mind when Fief supports multiple OAuth providers per user, thanks!
